contracts/facets/mmsa/MMSAFacet.sol

```diff
@@ -0,0 +1,433 @@
+import {IMMSAFacet} from "./interfaces/IMMSAFacet.sol";
+
+/**
+ * @title MMSA(Minimal Modular Smart Account) Facet
+ * @dev Facet enabling Barz to be ERC-7579 compatible account
+ * @author David Yongjun Kim (@PowerStream3604)
+ */
+contract MMSAFacet is
+    IMMSAFacet,
+    IAccount,
+    ModuleManager,
+    ValidationManager,
+    Executor,
+    BarzStorage
+{
+    using LibSentinelList for LibSentinelList.SentinelList;
+    using LibMode for ExecMode;
+
+    uint256 private constant VALIDATION_FAILED = 1;
+    address private constant ENTRYPOINT =
+        0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789; // v0.6 EntryPoint contract
+
+    // keccak256("Barz7579Message(bytes message)")
+    bytes32 private constant MMSA_MSG_TYPEHASH =
+        0x1efeaafa8dd3a6b211b37d11cde4213d4845ad021eed8dda5ba6003c5c9aa534;
+
+    modifier onlyEntryPointOrSelf() {
+        if (msg.sender != ENTRYPOINT && msg.sender != address(this)) {
+            revert MMSAFacet__NotfromEntryPointOrSelf();
+        }
+        _;
+    }
+
+    modifier withHook() {
+        address hook = address(LibMMSAStorage.mmsaStorage().hook);
+        if (hook == address(0)) {
+            _;
+        } else {
+            bytes memory hookData = IHook(hook).preCheck(
+                msg.sender,
+                msg.value,
+                msg.data
+            );
+            _;
+            IHook(hook).postCheck(hookData);
+        }
+    }
+
+    function initMMSA() external onlyEntryPointOrSelf {
+        _initialize();
+    }
+
+    function validateUserOp(
+        UserOperation calldata _userOp,
+        bytes32 _userOpHash,
+        uint256 /* missingAccountFunds */
+    ) external override returns (uint256 validationData) {
+        if (msg.sender != ENTRYPOINT) {
+            revert MMSAFacet__NotFromEntryPoint();
+        }
+
+        (, ValidationId validationId) = decodeNonce(_userOp.nonce);
+        validationData = _validate(validationId, _userOp, _userOpHash);
+    }
+
+    // UserOp.nonce is 32bytes in total
+    // |      1byte      |       20 bytes        | 1 byte |      2 bytes      |   8 bytes   |
+    // | Validation Type |    Validator Address  | unused |  validator system |   nonceKey  | -> If Validation Type is Validator
+    // |             Validation ID               |
+
+    // |      1byte      |  4 bytes   | 16 bytes | 1 byte |      2 bytes      |   8 bytes   |
+    // | Validation Type |PermissionId|  unused  | unused |  validator system |   nonceKey  | -> If Validation Type is Permission
+    // |             Validation ID               |
+    function decodeNonce(uint256 _nonce) internal pure returns (ValidationType validationType, ValidationId validation) {
+        assembly {
+            validationType := _nonce
+            validation := _nonce
+        }
+    }
+
+    function execute(
+        ExecMode _mode,
+        bytes calldata _executionCalldata
+    ) external payable override onlyEntryPointOrSelf onlyWhenUnlocked withHook {
```

Comment:
allowing self-call in execute functions should be considered carefully.

there is an attack scenario here, where a user is tricked to permit a session key access to address(this).execute(), not knowing what this permission really entails.

the session key can then build a userop. that has userop.calldata with two nested executions, the first one passing the sessionkey permission, the 2nd one would be arbitrary executions. this can lead to loss of funds

Reply:
@zeroknots Actually, the execute function of the 7579 part, the MMSAFacet's `execute()` cannot make a self call.
Self calls can only be made when the user uses the default VerificationFacet for UserOp Validation and use `executeSingle()` / `executeBatch()` from AccountFacetV2 for execution.

But it's a good callout, something worth deep consideration.